datamol/fp.py

```diff
@@ -3,59 +3,152 @@
 
 from rdkit import Chem
 from rdkit.Chem import rdMolDescriptors
-from rdkit.Chem import DataStructs
+from rdkit.Chem import rdmolops
+from rdkit.Chem import rdReducedGraphs
+from rdkit.Chem.EState import Fingerprinter as EStateFingerprinter
+from rdkit.Avalon import pyAvalonTools
+
+from rdkit.DataStructs.cDataStructs import ConvertToNumpyArray
+from rdkit.DataStructs.cDataStructs import SparseBitVect
+from rdkit.DataStructs.cDataStructs import UIntSparseIntVect
+from rdkit.DataStructs.cDataStructs import IntSparseIntVect
+from rdkit.DataStructs.cDataStructs import LongSparseIntVect
+from rdkit.DataStructs.cDataStructs import ExplicitBitVect
+from rdkit.DataStructs.cDataStructs import ULongSparseIntVect
 
 import numpy as np
 
 import datamol as dm
 
+_FP_FUNCS = {
+    "maccs": rdMolDescriptors.GetMACCSKeysFingerprint,
+    "ecfp": rdMolDescriptors.GetMorganFingerprintAsBitVect,
+    "topological": rdMolDescriptors.GetHashedTopologicalTorsionFingerprintAsBitVect,
+    "atompair": rdMolDescriptors.GetHashedAtomPairFingerprintAsBitVect,
+    "rdkit": rdmolops.RDKFingerprint,
+    "pattern": rdmolops.PatternFingerprint,
+    "layered": rdmolops.LayeredFingerprint,
+    "erg": rdReducedGraphs.GetErGFingerprint,
+    # NOTE(hadim): bad for pickling?
+    "estate": lambda x, **args: EStateFingerprinter.FingerprintMol(x)[0],
+    "avalon-count": pyAvalonTools.GetAvalonCountFP,
+    # NOTE(hadim): `rdkit-count` to ndarray hangs forever on my machine. Disabling it for now.
+    # "rdkit-count": rdmolops.UnfoldedRDKFingerprintCountBased,
+    "ecfp-count": rdMolDescriptors.GetHashedMorganFingerprint,
+    "fcfp-count": rdMolDescriptors.GetHashedMorganFingerprint,
+    "topological-count": rdMolDescriptors.GetHashedTopologicalTorsionFingerprint,
+    "atompair-count": rdMolDescriptors.GetHashedAtomPairFingerprint,
+}
+
 
-def fp_to_array(fp: DataStructs.ExplicitBitVect, dtype: type = int) -> np.ndarray:
+def fp_to_array(
+    fp: Union[np.ndarray, SparseBitVect, ExplicitBitVect, UIntSparseIntVect]
+) -> np.ndarray:
     """Convert rdkit fingerprint to numpy array.
 
     Note:
         This implementation has shown to be faster than using `DataStructs.ConvertToNumpyArray`
-        by a factor of ~4.
+        by a factor of ~4. See https://github.com/rdkit/rdkit/discussions/3863.
+
+    Args:
+        fp: The fingerprint.
     """
+
     if isinstance(fp, np.ndarray):
-        return fp
-    return np.frombuffer(fp.ToBitString().encode(), "u1") - ord("0")
+        fp_out = fp
+
+    elif isinstance(fp, SparseBitVect):
+        tmp = np.zeros(fp.GetNumBits(), dtype=int)
+        for n_bit in list(fp.GetOnBits()):
+            tmp[n_bit] = 1
+        fp_out = tmp
```

Comment:
Is that very efficient? Wondering whether `tmp[list(fp.GetOnBits())] = 1` would not be best here.